### ISSUE TITLE
Some caching occurs when using pylint in a server like mode.

### ISSUE THREAD

**Issue by ChienTeLee** [microsoft/vscode-pylint#129]
Hi, I just installed this extension and work on python 3.8.11. When I use relative import to import a file which have the same name with another file in other folder, pylint cannot import correctly and showing false positives. This happens globally, and I think it is the same issue with this one: https://github.com/PyCQA/pylint/issues/5645.

**Comment by karthiknadig** [comment]
@ChienTeLee This extension shows whatever we get from `pylint`. Try running pylint from the terminal and see if you get different results.

**Comment by DanielNoord** [comment]
I tried to create a reproducer for #132, but found that the extensions creates a different result compared to a manual invocation. I think it might be related to the bug reported here, but I'm not sure.

For:
```console
❯ cat test.py
def func(param: int) -> int:
    return param
❯ cat test2.py
import test

test.func(1)
```

`Pylint output`:
```
/Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/bin/python -m pylint --reports=n --output-format=json --from-stdin /Users/daniel/DocumentenLaptop/Programming/Github/pylint/test2.py
CWD Linter: /Users/daniel/DocumentenLaptop/Programming/Github/pylint
file:///Users/daniel/DocumentenLaptop/Programming/Github/pylint/test2.py :
[
    {
        "type": "error",
        "module": "test2",
        "obj": "",
        "line": 3,
        "column": 0,
        "endLine": 3,
        "endColumn": 9,
        "path": "test2.py",
        "symbol": "no-member",
        "message": "Module 'test' has no 'func' member",
        "message-id": "E1101"
    }
]

```

`Manual`:
```
❯ /Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/bin/python -m pylint --reports=n --output-format=json /Users/daniel/DocumentenLaptop/Programming/Github/pylint/test2.py
[]
```

Note that I removed ``--from-stdin``. I think we looked into `--from-stdin` previously, but there might still be something broken about that flag.

**Comment by karthiknadig** [comment]
@DanielNoord Can you dump what `pylint` sees as sys path in the manual run case. I think that might be the problem. Extension is not setting up the `sys.path`, as `pylint` would expect to see it.

**Comment by DanielNoord** [comment]
```python
❯ cat test2.py | /Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/bin/python -m pylint --reports=n --output-format=json --from-stdin /Users/daniel/DocumentenLaptop/Programming/Github/pylint/test2.py
['/Users/daniel/DocumentenLaptop/Programming/Github/pylint',
'/Users/daniel/.pyenv/versions/3.10.0/lib/python310.zip',
'/Users/daniel/.pyenv/versions/3.10.0/lib/python3.10', 
'/Users/daniel/.pyenv/versions/3.10.0/lib/python3.10/lib-dynload',
'/Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/lib/python3.10/site-packages',
'/Users/daniel/DocumentenLaptop/Programming/Github/pylint',
'/Users/daniel/DocumentenLaptop/Programming/Github/astroid']
[]
```

Just before we actually start checking the file.

--------------------

```diff
diff --git a/pylint/lint/pylinter.py b/pylint/lint/pylinter.py
index e2bbf061e..092bdd6f9 100644
--- a/pylint/lint/pylinter.py
+++ b/pylint/lint/pylinter.py
@@ -715,6 +715,9 @@ class PyLinter(
         :param FileItem file: data about the file
         :raises AstroidError: for any failures stemming from astroid
         """
+        import sys
+
+        print(sys.path)
         self.set_current_module(file.name, file.filepath)
         # get the module representation
         ast_node = get_ast(file.filepath, file.name)
```

**Comment by DanielNoord** [comment]
`sys.path` of extension:
```python
   /Users/daniel/.vscode/extensions/ms-python.pylint-2022.3.11671003/bundled/linter
   /Users/daniel/.pyenv/versions/3.10.0/lib/python310.zip
   /Users/daniel/.pyenv/versions/3.10.0/lib/python3.10
   /Users/daniel/.pyenv/versions/3.10.0/lib/python3.10/lib-dynload
   /Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/lib/python3.10/site-packages
   /Users/daniel/DocumentenLaptop/Programming/Github/pylint
   /Users/daniel/DocumentenLaptop/Programming/Github/astroid
   /Users/daniel/.vscode/extensions/ms-python.pylint-2022.3.11671003/bundled/libs
```

@karthiknadig How can I read anything printed to `sys.stdout` from the Extension?

**Comment by DanielNoord** [comment]
I noticed something strange as well. I'm using an editable instal `pip install -e .` in my environment. The extensions picks this up and shows that it is using both my local version of `astroid` (our underlying interpreter) and `pylint`.

However, whenever I make changes to those installs they don't get used by the extension. I need to restart the server for those changes to take effect.
So, when I tried to print out `sys.path` using the `diff` above for what the extensions uses (so not what the extension reports, but what `pylint` actually sees during a run) I had to restart the server for the print statement to take effect. So some element of the linter is being kept around in between server runs. This could create a "cache" that `pylint` doesn't know how to handle.

Furthermore, the `sys.path` reported by the extension and the `sys.path` used by `pylint` when run from the extension are not the same. For the `sys.path` above, using the diff from two posts previous I actually get:
```python
['/Users/daniel/DocumentenLaptop/Programming/Github/pylint',
'/Users/daniel/.pyenv/versions/3.10.0/lib/python3.10',
'/Users/daniel/.pyenv/versions/3.10.0/lib/python3.10/lib-dynload',
'/Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/lib/python3.10/site-packages', 
/Users/daniel/DocumentenLaptop/Programming/Github/astroid',
'/Users/daniel/DocumentenLaptop/Programming/Github/pylint',
'/Users/daniel/.vscode/extensions/ms-python.pylint-2022.3.11671003/bundled/libs']
```

This could be because of the changes to `sys.path` `pylint` makes, but I think it might be useful to know that these are not the same.

**Comment by karthiknadig** [comment]
@DanielNoord for diagnostic purposes, I dump everything that i get from `sys.stdout` each time to the Output>pylint. what you see in the logs when the linter runs on a file is exactly what we got from `sys.stdout`. `result.stdout` is the content of the substituted stream for stdout.
https://github.com/microsoft/vscode-pylint/blob/9d05b8efd44d6aa9b9fd74cff18e4854c13b85b2/bundled/linter/linter_server.py#L258

If you want verbose logging then add `"pylint.trace.server": "verbose",` to your VSCode user (not workspace) settings.

**Comment by karthiknadig** [comment]
> So, when I tried to print out sys.path using the diff above for what the extensions uses (so not what the extension reports, but what pylint actually sees during a run) I had to restart the server for the print statement to take effect. So some element of the linter is being kept around in between server runs. This could create a "cache" that pylint doesn't know how to handle.

I believe this might be due to `runpy` and extension never unloading `pylint` after a running. The extension, loads pylint once, and reuses it. In the normal scenario, `pylint` would run, and exit, and the only way some state was persisted was by writing to a file. Now that pylint runs like a server, it seems some state is being re-used.

**Comment by DanielNoord** [comment]
Yeah, some of the caching in `astroid` depends on global states so that is likely causing issues then. I think it would take quite some time before we can untangle all the different states `pylint` and `astroid` need/use before we can smoothly support a server-like run. Is there anything we can do with `runpy` to avoid this? Like passing a flag to re-load?

**Comment by karthiknadig** [comment]
@brettcannon Any suggestions here on how we can do reload safely?

**Comment by karthiknadig** [comment]
For completeness I captured the sys.path for the ways of running pylint that we were looking at:

### Running pylint as a module in terminal:
```
>type test2.py | python -m pylint --reports=n --output-format=json --from-stdin test2.py
sys.path = [
    'C:\\GIT\\repro\\linter2',
    'C:\\all_pys\\Python37\\python37.zip',
    'C:\\all_pys\\Python37\\DLLs',
    'C:\\all_pys\\Python37\\lib',
    'C:\\all_pys\\Python37',
    'C:\\GIT\\repro\\linter2\\.venv',
    'C:\\GIT\\repro\\linter2\\.venv\\lib\\site-packages'
]
```

### Using runpy.run_module

```
>type test2.py | python viarunpy.py --reports=n --output-format=json --from-stdin test2.py
sys.path = [
    'C:\\GIT\\repro\\linter2',
    'C:\\all_pys\\Python37\\python37.zip',
    'C:\\all_pys\\Python37\\DLLs',
    'C:\\all_pys\\Python37\\lib',
    'C:\\all_pys\\Python37',
    'C:\\GIT\\repro\\linter2\\.venv',
    'C:\\GIT\\repro\\linter2\\.venv\\lib\\site-packages'
]
```

### Via the extension

The typescript part launches this `python bundled/linter/linter_server.py`. Pylint is eventually launched by the handler for file open using `runpy.run_module`.
```
sys.path = [
    'c:\\GIT\\repro\\linter2',
    'C:\\all_pys\\Python37\\DLLs',
    'C:\\all_pys\\Python37\\lib',
    'C:\\all_pys\\Python37',
    'c:\\GIT\\repro\\linter2\\.venv',
    'c:\\GIT\\repro\\linter2\\.venv\\lib\\site-packages',
    'c:\\GIT\\linters\\vscode-pylint\\bundled\\libs'
]
```

The only thing missing from the sys.path in the extension `C:\\all_pys\\Python37\\python37.zip`, not sure why that gets skipped. May be because it is not running in a terminal?

**Comment by DanielNoord** [comment]
@jacobtylerwalls You have a better understanding of clearing caches within `astroid` than I do. Do you think there is something we can easily fix before we eventually go through the effort of allowing a server-like pylint process to run?

**Comment by jacobtylerwalls** [comment]
> Any suggestions here on how we can do reload safely?

Should be as simple as calling `clear_cache()` on any `AstroidManager` object, which can be accessed as either `astroid.MANAGER` or `pylint.lint.pylinter.MANAGER`. See the recent (and well-hidden) documentation at the bottom of this page:

https://pylint.pycqa.org/en/latest/development_guide/api/pylint.html?highlight=between%20runs#edited-between-runs

This should be done after every pylint call. Can we try that here?

If _that_ doesn't work, then there's still a global state somewhere we don't know about (or we need to look into Brett's tip about importing off the module and verify it's not that). This is all assuming a single-process scenario. There are many known issues with parallel mode in `pylint` and it's not really safe to use IMO.

***
@DanielNoord we could look into some usability improvements, either always clearing the cache at the end of a run, and/or providing a kwarg argument to opt _in_ to retaining the cache. WDYT?

**Comment by DanielNoord** [comment]
@jacobtylerwalls Seems like an option ``--clear-cache-post-run`` (forgive the name) would indeed be useful. That would require some version checking for this plugin but might be more user friendly going forward.

**Comment by brettcannon** [comment]
> This should be done after every pylint call. Can we try that here?

Yep, that should be doable. We can probably handle it somewhere in https://github.com/microsoft/vscode-pylint/blob/9d05b8efd44d6aa9b9fd74cff18e4854c13b85b2/bundled/linter/utils.py#L112-L134 .



> There are many known issues with parallel mode in `pylint` and it's not really safe to use IMO.

We are running single-threaded, so that's not a concern here (perk of Pylint being run in a language server and those out-of-process to VS Code itself).

**Comment by bjtho08** [comment]
Is this still being investigated? I have the same issue when doing refactoring in a project. Every time I change the name of a class, function or method that is imported elsewhere, I need to reload pylint to clear the error that is sure to show up after I save the changes, which triggers linting of the open files.

**Comment by brettcannon** [comment]
@DanielNoord did any flag ever end up in Pylint? Or should we just hack up a forced reload of the appropriate modules?

**Comment by jacobtylerwalls** [comment]
I put https://github.com/PyCQA/pylint/pull/7802 up just now as a proof of concept. I'd like to come back and add a test.

**Comment by DanielNoord** [comment]
Does anybody from the team want to try out the new flag? We could put it an a patch release and get it out relatively easily, but since your the main target for this flag we'd like to get confirmation that we did indeed fix the issue. If not, we could try and find something else!

**Comment by brettcannon** [comment]
We might not have time to get to this until January, but we will definitely give it a try when we can! No need to do a point release just for us.

**Comment by DanielNoord** [comment]
Thanks for getting back! I'll guess we'll just merge then and release at some point before January. We think this fits the use case, but don't hesitate to tell us if we missed anything!

**Comment by karthiknadig** [comment]
@DanielNoord Which version of pylint has the flag?

**Comment by jacobtylerwalls** [comment]
2.16, which hasn't been released yet. See PyCQA/pylint#7802

**Comment by Pierre-Sassoulas** [comment]
We're planning on releasing 2.16.0 *soonish* :tm:
